app/src/main/java/org/lineageos/twelve/viewmodels/VisualizerViewModel.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-FileCopyrightText: 2024 The LineageOS Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.lineageos.twelve.viewmodels
+
+import android.app.Application
+import android.media.audiofx.Visualizer
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.flow.stateIn
+import org.lineageos.twelve.TwelveApplication
+import org.lineageos.twelve.ext.waveFormFlow
+
+class VisualizerViewModel(application: Application) : AndroidViewModel(application) {
+    private val samplingRate = Visualizer.getMaxCaptureRate()
+
+    private val visualizerFlow = channelFlow {
+        val visualizer = Visualizer(getApplication<TwelveApplication>().audioSessionId)
+        visualizer.setEnabled(true)
+
+        trySend(visualizer)
+
+        awaitClose {
+            visualizer.setEnabled(false)
+            visualizer.release()
+        }
+    }
+
+    private val visualizer = visualizerFlow
+        .flowOn(Dispatchers.Default)
+        .stateIn(
+            viewModelScope,
+            started = SharingStarted.WhileSubscribed(),
+            initialValue = null
+        )
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val waveform = visualizer.filterNotNull()
+        .flowOn(Dispatchers.Default)
+        .flatMapLatest { it.waveFormFlow(samplingRate) }
+        .stateIn(
+            viewModelScope,
+            started = SharingStarted.WhileSubscribed(),
+            initialValue = ByteArray(0)
+        )
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val amplitude = waveform
+        .mapLatest {
+            it.maxOrNull() ?: 0
```

Comment:
I think this is wrong, prolly needs the same toSignedShort dance to normalize it